src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
+    // target or the stage, or null if no matching list is found.
+
+    let list = target.getList(listName);
+    list = list || stage.getList(listName);
+    return list ? list.id : null;
+  }
+
+  interface SerializeInputsToFields {
+    stage: Stage;
+    target: Target;
 
-function extractCostumes(target: Target): sb3.Costume[] {
-  return target.costumes.map(costume => {
-    const sb3Costume: sb3.Costume = {
-      assetId: costume.md5,
-      name: costume.name,
-      md5ext: `${costume.md5}.${costume.ext}`,
-      dataFormat: costume.ext,
-      rotationCenterX: costume.centerX / 2,
-      rotationCenterY: costume.centerY / 2
+    fieldEntries;
+  }
+
+  function serializeInputsToFields(
+    inputs: { [key: string]: BlockInput.Any },
+    options: SerializeInputsToFields
+  ): {
+    fields: { [key: string]: any };
+  } {
+    // Serialize provided inputs into a "fields" mapping that can be stored
+    // on a serialized block.
+    //
+    // Where the Scratch 3.0 term "input" refers to a slot that accepts blocks,
+    // the term "field" is any user-interactive slot that cannot be obscured as
+    // such. Fields are also the interactive element within any shadow block,
+    // making their presence key to nearly all inputs.
+    //
+    // While inputs are fairly complex to serialize, fields are comparatively
+    // simple. A field always contains only its selected or entered value, so
+    // it's not concerned with a variety of structures like an input is.
+    // The format for a field mapping looks something like this:
+    //
+    //   {
+    //     VARIABLE: ["my variable", "theVariablesId"]
+    //   }
+    //
+    // Some fields take an ID; some don't. Here's another example:
+    //
+    //   {
+    //     DISTANCETOMENU: ["_mouse_"]
+    //   }
+    //
+    // The important thing to remember with fields during serialization is that
+    // they refer to slots that don't accept blocks: non-droppable menus,
+    // primarily, but the value in a (non-compressed) shadow input too.
+
+    const { fieldEntries, stage, target } = options;
+
+    const result = {
+      fields: {}
     };
-    return sb3Costume;
-  });
-}
 
-function extractSounds(target: Target): sb3.Sound[] {
-  return target.sounds.map(sound => {
-    const sb3Sound: sb3.Sound = {
-      assetId: sound.md5,
-      name: sound.name,
-      dataFormat: sound.ext,
-      format: "",
-      rate: sound.sampleRate,
-      sampleCount: sound.sampleCount,
-      md5ext: `${sound.md5}.${sound.ext}`
+    if (!fieldEntries) {
+      return result;
+    }
+
+    for (const key of Object.keys(fieldEntries)) {
+      const input = inputs[key];
+      // Fields are stored as a plain [value, id?] pair.
+      let id: string;
+      switch (input.type) {
+        case "variable":
+          id = getVariableId(input.value, target, stage);
+        case "list":
+          id = getListId(input.value, target, stage);
+        default:
+          id = null;
```

Comment:
```suggestion
        case "variable":
          id = getVariableId(input.value, target, stage);
          break;
        case "list":
          id = getListId(input.value, target, stage);
          break;
        default:
          id = null;
          break;
```
Watch out for fallthrough!